cmd/sprint-automation/main.go

```diff
@@ -386,6 +447,112 @@ func sendIntakeDigest(slackClient *slack.Client, jiraClient *jiraapi.Client, use
 	return nil
 }
 
+type versionInfo struct {
+	stable          bool
+	version         string
+	state           configv1.UpdateState
+	semanticVersion semver.Version
+}
+
+func newVersionInfo(status configv1.ClusterVersionStatus) (*versionInfo, error) {
+	if len(status.History) == 0 {
+		return nil, fmt.Errorf("failed to get history of ClusterVersion version")
+	}
+	current := status.History[0]
+	ret := &versionInfo{
+		version: current.Version,
+		state:   current.State,
+		stable:  current.State == configv1.CompletedUpdate && current.CompletionTime != nil && time.Since(current.CompletionTime.Time) > 24*time.Hour,
+	}
+	cv, err := semver.Make(current.Version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine sematic version: %s", current.Version)
+	}
+	ret.semanticVersion = cv
+	if ret.stable && len(status.History) > 1 {
+		previous := status.History[1]
+		pv, err := semver.Make(previous.Version)
+		if err != nil {
+			return nil, fmt.Errorf("failed to determine sematic version: %s", previous.Version)
+		}
+		if cv.Minor > pv.Minor {
+			// Need to soak a week after a Y-stream upgrade
+			ret.stable = time.Since(current.CompletionTime.Time) > 7*24*time.Hour
+		}
```

Comment:
Is there a window for this to actually fire without racing? We upgrade b01 weekly automatically, at the same time.

So b01 starts upgrading at Week 1, Thursday 10:00 (just an example). It finishes several hours later, at 15:00. So this condition will hold on Week 2 Thursday 15:00. But at that time b01 will be already upgraded to next version (because it will start upgrading on Week 2 Thursday 10:00).

Reply:
Ah right, our prowjob runs every day, but OCP zstreams are basically weeklies.

Yeah, I'm aware of the safeguard - I'm not worried that it will instruct triage to upgrade to a wrong version, I'm worried that it will not instruct triage to update at all, because it seems that b01 is rarely on a single version for more than 7 days. Basically after 7 days b01 will often start upgrading to a new version.

Although... it seems that there *are* windows. Maybe shorten the interval to six days (instead of seven)? That means that if the releases go out weekly, we will always upgrade b02 a day before b01 starts upgrading to the next version.

![b01](https://user-images.githubusercontent.com/712614/142905858-156fcb4f-f5cf-48f7-9694-7cc0628d180e.png)

Reply:
>I'm worried that it will not instruct triage to update at all, because it seems that b01 is rarely on a single version for more than 7 days.

A 7-day soaking is a requirement only for Y-stream upgrade.
We usually have a subtask while dong it.
https://issues.redhat.com/browse/DPTP-2504

E.g., 

```json
  {
    "completionTime": "2021-11-11T08:37:17Z",
    "image": "quay.io/openshift-release-dev/ocp-release@sha256:5c55be02e32e688ec5a404858a08cf533ba15b50b6f0e028089635b47db5866e",
    "startedTime": "2021-11-11T07:30:55Z",
    "state": "Completed",
    "verified": true,
    "version": "4.9.7"
  },
  {
    "completionTime": "2021-11-01T14:06:52Z",
    "image": "quay.io/openshift-release-dev/ocp-release@sha256:386f4e08c48d01e0c73d294a88bb64fac3284d1d16a5b8938deb3b8699825a88",
    "startedTime": "2021-11-01T12:47:30Z",
    "state": "Completed",
    "verified": true,
    "version": "4.9.5"
  },
```

b01 stayed with 4.9.5 (which is a Y stream upgrade from 4.8.17) from "2021-11-01T14:06:52Z" to "2021-11-11T07:30:55Z".
Over 7 days in total.
Sprint automation will start to ping triage from "2021-11-08T14:06:52Z".

Z stream upgrades require only 1 day of soaking.

Reply:
OK I get it now, sorry :facepalm: 

So, for a future me who asks dumb questions, can we add a comment that describes what the logic intends to do? Ping triage about zstream after a day of soaking, ping triage about y stream after a week of soaking? The logic is clear to me now when you explained it to me but... I needed that explanation ;)

Reply:
Added comments.
Sorry about the confusion.